### PR TITLE
Fixed #31546 -- Deprecated boolean value for Command.requires_system_checks.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/31546

### CODE REVIEW

File: django/core/management/base.py [hunk 1]
```diff
@@ -365,7 +374,11 @@ def execute(self, *args, **options):
             self.stderr = OutputWrapper(options['stderr'])
 
         if self.requires_system_checks and not options['skip_checks']:
-            self.check()
+            tags = None
+            if isinstance(self.requires_system_checks, list):
+                tags = self.requires_system_checks
+            self.check(tags=tags)
```
Comment (felixxm) [inline django/core/management/base.py:380]: We should validate `self.requires_system_checks`, currently we run all checks also for any truthy value e.g. `requires_system_checks = 'x'`, maybe
```python
if not isinstance(value, (list, tuple)) and value != '__all__':
    raise TypeError(...)
```

File: django/core/management/base.py [hunk 2]
```diff
@@ -240,6 +241,14 @@ def __init__(self, stdout=None, stderr=None, no_color=False, force_color=False):
             self.style = color_style(force_color)
             self.stderr.style_func = self.style.ERROR
 
+        if self.requires_system_checks in [False, True]:
+            warnings.warn(
+                "Using 'requires_system_checks = %r' is deprecated, "
+                "use 'requires_system_checks = %r' instead." %
+                (self.requires_system_checks, '__all__' if self.requires_system_checks else []),
+                RemovedInDjango40Warning,
+            )
```
Comment (felixxm) [inline django/core/management/base.py:250]: We should update `self.requires_system_checks` if it's a boolean, e.g.
```python
self.requires_system_checks = '__all__' if self.requires_system_checks else []
```
we can also add a module level constant:
```
ALL_CHECKS = '__all__'
```

File: docs/releases/3.2.txt [hunk 3]
```diff
@@ -253,5 +253,10 @@ Features deprecated in 3.2
 
 Miscellaneous
 -------------
+* Using boolean value in :attr:`BaseCommand.requires_system_checks
+  <django.core.management.BaseCommand.requires_system_checks>` is deprecated.
+  Use ``'__all__'`` instead of ``True``, ``[]`` instead of ``False``, and a
+  list of :ref:`Django’s system checks tags <system-check-builtin-tags>` to
+  validate some application's models rather than all applications' models.
```
Comment (felixxm) [inline docs/releases/3.2.txt:260]: Please replace `* ...`

File: docs/internals/deprecation.txt [hunk 4]
```diff
@@ -103,6 +103,9 @@ details on these changes.
 See the :ref:`Django 3.1 release notes <deprecated-features-3.1>` for more
 details on these changes.
 
+* The ability to use boolean value in :attr:`BaseCommand.requires_system_checks
+  <django.core.management.BaseCommand.requires_system_checks>` will be removed.
```
Comment (felixxm) [inline docs/internals/deprecation.txt:107]: This should be moved to features removed in Django 4.1.

File: docs/howto/custom-management-commands.txt [hunk 5]
```diff
@@ -215,8 +215,13 @@ All attributes can be set in your derived class and can be used in
...
 
-    A boolean; if ``True``, the entire Django project will be checked for
-    potential problems prior to executing the command. Default value is ``True``.
+    A string or list; if ``'__all__'``, the entire Django project will be
+    checked for errors prior to executing the command. Default value
+    is ``'__all__'``. To validate some application's models rather than
+    all application's models, change it by a list of
+    :ref:`Django’s system checks tags <system-check-builtin-tags>`:
+
+    ``requires_system_checks = [Tags.staticfiles, Tags.models]``.
```
Comment (felixxm) [inline docs/howto/custom-management-commands.txt:224]: `versionchanged` is missing.

Comment (hramezani) [reply docs/howto/custom-management-commands.txt:224]: Added!

File: docs/internals/deprecation.txt (hunk 4 above)
Comment (hramezani) [reply docs/internals/deprecation.txt]: I moved it to features removed in Django 4.1 and also replaces all `RemovedInDjango40Warning` with `RemovedInDjango41Warning`

File: docs/releases/3.2.txt (hunk 3 above)
Comment (hramezani) [reply docs/releases/3.2.txt]: It replaced in another commit.

File: django/core/management/base.py (hunk 1 above)
Comment (hramezani) [reply django/core/management/base.py:391]: Validation added and a test added to check raising of the `TypeError`

File: django/core/management/base.py (hunk 2 above)
Comment (hramezani) [reply django/core/management/base.py:252]: Updated!

File: tests/admin_scripts/tests.py [hunk 6]
```diff
@@ -1791,6 +1794,117 @@ def test_label_command_multiple_label(self):
             "('settings', None), ('traceback', False), ('verbosity', 1)]"
         )
 
+    def test_requires_system_checks_empty_list(self):
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:1797]: Moved to `user_commands.tests`.

File: tests/admin_scripts/tests.py [hunk 7]
```diff
@@ -1791,6 +1794,117 @@ def test_label_command_multiple_label(self):
...
+        out = StringIO()
+        command = Command(stdout=out)
+        with mock.patch('django.core.management.base.BaseCommand.check') as mocked_check:
+            call_command(command, skip_checks=False)
+        self.assertEqual(out.getvalue(), "Hello, World!\n")
+        self.assertTrue(mocked_check.called)
+        _, kwargs = mocked_check.call_args
+        self.assertEqual(kwargs, {'tags': None})
+
+    def test_requires_system_checks_tags_list(self):
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:1827]: Moved to `user_commands.tests`.

File: tests/admin_scripts/tests.py [hunk 8]
```diff
@@ -1791,6 +1794,117 @@ def test_label_command_multiple_label(self):
...
+        out = StringIO()
+        command = Command(stdout=out)
+        with mock.patch('django.core.management.base.BaseCommand.check') as mocked_check:
+            call_command(command, skip_checks=False)
+        self.assertEqual(out.getvalue(), "Hello, World!\n")
+        self.assertTrue(mocked_check.called)
+        _, kwargs = mocked_check.call_args
+        self.assertEqual(kwargs, {'tags': ['staticfiles', 'models']})
+
+    def test_requires_system_checks_invalid_value(self):
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:1843]: Moved to `user_commands.tests`.

**Pull request merged**